### ISSUE TITLE
handling both hash routing & anchor links

### ISSUE THREAD

**Issue by sethvincent** [choojs/choo#440]
As I'm porting [minidocs to choo v5](https://github.com/freeman-lab/minidocs/tree/choo-v5) I'm working on making choo v5 support both hash routing & navigating to anchor links if they are present on the page.

The tricky part is that when the router is first called, the element isn't available on the page yet: https://github.com/yoshuawuyts/choo/blob/master/index.js#L50

So there's not a way to check if an id is available to scroll to before the router is called.

Thoughts on how to implement this?

**Comment by sethvincent** [comment]
Forgot to add:

Current behavior of `createLocation` function is to rewrite a url fragment like `/about#install` to `/about/install` before it is passed to the router.

The goal (I think) is to avoid this behavior when the `#install` id exists on the page, and only tell the router the pathname in that case.

**Comment by jenslind** [comment]
Instead of checking if the id exists, perhaps we could prefix the hash if we want to use it as an anchor. Something like this: `/route#anchor:id`, just a thought.

**Comment by yoshuawuyts** [comment]
Heya, thanks for the suggestion! - Personally I wouldn't want a framework
to tell me how to write my links, but just work as expected. Think we
should explore making that work first (think it might already? not sure,
not tested) and only if that's truly impossible look for different options.
Hope that makes sense ✨

**Comment by jenslind** [comment]
Totally agree!

Another thought off the top of my head.. 💭 
Having `#` behaving differently depending on page content feels a bit weird to me.
What about an option to turn off support for `#` in routes, i.e. do not turn `/some#route` into `/some/route`. And let the user make the choice of treating `#` as a path or anchor.

🐈

**Comment by yoshuawuyts** [comment]
@jenslind generally I'm not a big fan of options, but as per #489 by exposing the `app._createRoute` method the behavior can be changed if people wish to. Hope this good enough! :D

**Comment by hbrls** [comment]
Do you have a hash routing example? I cannot get it to work.

**Comment by bcomnes** [comment]
Simple example https://github.com/hypermodules/hyperamp/blob/master/renderer/index.js

**Comment by LJNGDAHL** [comment]
As a fairly new user of choo (and so far overall super impressed by it 🎉) I’m having a bit of trouble understanding the way anchor links are handled. How come you have decided to turn `/some#route` into `/some/route`?

It took awhile for me to wrap my head around what the section [Hash routing](https://github.com/choojs/choo#hash-routing) in the README.md really meant (and how to get anchor links working like they usually do) – but that might just be me as a confused newbie. 🤔 ✨

**Comment by johnelliott** [comment]
It seems like the always-there default to <a href="#">top</a> is does not with by default with nanohref. It would be nice if that worked unless a route is set on #.

**Comment by earthtone** [comment]
Been experimenting with migrating over a legacy project over to choo, and I came across a syntactical particular that may need documenting. Like @LJNGDAHL I too was a bit confused by how to implement hash routing for my particular use case, which is slightly different from the one described in the README. I hope this is the right issue to report this on.

Previously I was using [yo-yo for components](https://github.com/maxogden/yo-yo) and an [unrelated library](https://github.com/krasimir/navigo) for hash routing, because the rest of my team felt it was easier to parse data about the user's activity using the hash as a sort-of namespace. So, for example, `hostname.com/#/library/book-name/book-id/view-name`. There are some cases within the app, where a specific instance of a dynamic URL (e.g. `#/library/moby-dick/12345/page/6`) will need to be linked to somewhere else in the app. In this case, it was decided that API would provide the specific URLs, which is why I'm now stuck with this particular naming scheme.

Defining hash routes with choo either as `app.route('#/library/:bookName/:bookId/page')` or `app.route('#library/:bookName/:bookId/page`) both resolve as `hostname.com/#library/book-name/book-id/page`. Because there is already an existing naming scheme I need to follow, that requires the `/` after the `#`, this didn't work for me. However, defining a route as `app.route('#//library/:bookName/:bookId/page')` resolves as `hostname.com/#/library/book-name/book-id/page`, which is what I was looking for.

**TL;DR**
Choo's router ignores the first `/` following a `#` in the route definition. This might be a minor syntactical detail but it's useful for dev's to be aware of it if they have particular URL needs/requirements on their projects, IMO.

**Comment by afincato** [comment]
`hash routing` works 'out of the box' except when you are at the root page and have a link bringing somewhere to that page — in which case a `/` is added before the `#link`.

how do you prevent that?

**Comment by istvanvasil** [comment]
>hash routing works 'out of the box' except when you are at the root page and have a link bringing somewhere to that page — in which case a / is added before the #link.

> how do you prevent that?

Any success on this?
